### ISSUE TITLE
[notifications]: no history for certain notifications (example cantata: "now playing" message bubble is shown, but not kept), those were stored for later reading in kde4.x

### ISSUE THREAD

**Issue by star-buck** [blue-systems/plasma-5.4#48]

**Comment by thugcee** [comment]
> Can you think of any other notifications besides media player ones that the user would want to keep in the history (and why)?

I can think of many such notifications. I have just switched to Plasma 5 and lack of notifications history is one of two biggest minuses compared to KDE4. 

example: I leave my computer to make a coffee or I watch a movie in full screen mode and then I go back to desktop and I want to know: 
- if any emails have arrived in mid time,
- if there was any bell on one of many of my console sessions,
- if backup script have finished successfully, 
- if log watcher have detected something unusual, 
- and I can go and go with such examples.

And you say that it's now better, that every time I stop watching desktop for a moment, I have to check all notification sources to make sure I have not missed anything. For me this is stupid.

I have switched to KDE because its developers, instead of forcing ideas, they were giving its users freedom. Please do not go the same way - "the one correct way for all" - Gnome and Unity have come.

**Comment by antonvinny** [comment]
Just migrated from KDE 4 and really missing this feature. An option to enable the notification history would be a great feature.

**Comment by Houston4444** [comment]
I really agree with thugcee. It should be really, really better to let user choose and keep an option to get a notifications history. And I didn't find any option to make a persistent notification with a shell script (Not possible with kdialog for example). If a script make new files and user don't see these files, now he will never find them.

**Comment by ManIVIctorious** [comment]
This is really annoying. I just missed 2 quite important eMails and lost about an hour searching for this feature, just to realize it has been removed. Pleas let the user decide if he/she wants the messages saved, my mail client (geary) obviously doesn't think that an eMail notification should be persistent => maybe the developer doesn't always know best...

**Comment by christophschw** [comment]
Yeah, I do agree. It missed a lot of messages from kdeconnect and from my mailnotifier because I wasn't in front of my PC. 

At the moment my notification icon is unused because there aren't any persistent notifications. Would be great to see a possibility to change the history behaviour. Thanks!

**Comment by konqoro** [comment]
Clementine used to pollute the notification tray with many "Now playing..." notifications (when configured to use native notifications) Thus this is an improvement, not a regression. Thank you.

**Comment by Houston4444** [comment]
It's maybe an improvement for music players softs which don't choose OSD. But Notifications ARE NOT ONLY DONE for music information, but for emails, little scripts, new files on Dropbox, and a lot of things you NEED to know. The choice of devs is understandable, Don't let user choose is a regression.

**Comment by thugcee** [comment]
@konqoro Then Clementine should be fixed (to use OSD for example) instead of breaking notifications system.

**Comment by sfrique** [comment]
I agree with @thugcee, i need ot know if i got an email or anything when i get back to computer..

**Comment by ManIVIctorious** [comment]
@konqoro just use clementines OSD!

**Comment by christophschw** [comment]
@thugcee You're completely right!

**Comment by wintericie** [comment]
+1 for this issue.

**Comment by frasty** [comment]
Breaking the notification system just to let some audio player not to pollute it doesn't sound as a good idea to me. I have yet to see an application who does use persistent notifications, thus as of now the notification system is quite useless..

**Comment by hetdegon** [comment]
I agree with @thugcee as well. Notification history is not very useful now. It's always empty.

**Comment by jayeshbadwaik** [comment]
While I agree that the default behavior should be to conform to the freedesktop standards. There should be an option by the user which overrides the default behvaior, simply because so many people want that option to be there.

**Comment by nortexoid** [comment]
Like many others, I don't wish the fate of my notifications to lie in the hands of random application developers! The user should have the option to keep a history of all notifications (or better, those she chooses). By the way, like KDE4, Gnome (3.18) DOES keep a history of all notifications, accessible by clicking on the time in the panel, with an option to dismiss all. Very useful!

**Comment by vindicatorr** [comment]
I want to +1 this as well.
I just moved from Ubuntu 15.10 to Arch with KDE, installed KDE Connect and while notifications DO pop up from my phone, sometimes I'm engrossed in what I'm doing that I don't see them in time, or like others say, I can easily be away from my computer and not know anything came up because it isn't "persistent" in the notification history.

+1 for "user's choice".
If, as a developer of this piece, you don't like the idea, just make the default choice to not retain all items in the history and only those with programs that enable persistence, but give the user that option to change it... unless you're just making this piece only for you...
But there are probably going to be cases of dead programs (\<cough\> skype) that won't even update with new system UI changes so notifications will never persist.

**Comment by colomar** [comment]
The KDE usability team has recognized that missing notifications is a problem, but we do not consider the behavior in Plasma 4 to be the optimal solution.
I will try to get in touch with former leader of the team Celeste again, who happens to have written her PhD thesis about the topic of desktop notifications. If there is any one person who knows best how to solve this issue, it's her.
I don't want to shoot a solution from the hip as that would be likely to please have of our users while annoying the other half, so please give us some time to come up with a good solution,

**Comment by star-buck** [comment]
@colomar: Thanks for giving it another shot.
How about taking a look at how others solve this, like Gnome, Mint, Deepin, or even recent Windows10.
As to technical how: A checkbox-option to keep messages present until manual discarded should solve the problem imo.

**Comment by colomar** [comment]
@star-buck Yes, looking at others will be part of the strategy. Getting scientific input will be another part, as Celeste already promised to look into it :)
Providing a config option is possible, but should only be our last resort. First we should see if we can find a solution that works well for everyone.

**Comment by mck182** [comment]
> Yes, looking at others will be part of the strategy

Fwiw, I did look at others a year ago, see [the comment at the top](https://github.com/blue-systems/plasma-5.4/issues/48#issuecomment-68208241). Will be interesting to see if any of those changed their implementation.

**Comment by sfrique** [comment]
@colomar I dont see whymaking a config should be the last resort.
You could define the default for "everyone" but could give the choice for
the people that don't want the defaut options.

The one thing KDE/PLASMA had it was that you could configure pretty much
anything. As gnome has to opposing view that is not able to configure it
all and use what they provide.

Thanks

Em qua, 3 de fev de 2016 às 15:57, mck182 <notifications@github.com>
escreveu:

**Comment by llelectronics** [comment]
Yeah I am also pro when it comes to make it configurable. 
The default should be like is (as this is basically the default on other desktops aswell. @mck182 basically nothing changed on the other desktops) 
If the applications don't make it easy to choose how their notifications are shown (persistent or not) then the desktop should allow such an option. 
It should be called "Save notification history" and have a combobox (the one with the numbers) where you can choose the maximum of notifications to be stored until the older ones get removed. 
Or it should be made configurable how long notifications in general are shown and then maybe stored for later reading. 
The overall best solution would be to have a global rules list where one can set which notifications to store persitent and which not.

**Comment by mck182** [comment]
As I said in my comments above - we have much better
tools to actually do what you want from persistent notifications.
It's called KStatusNotifierItem and it is sooo much better
and offers sooo much more possibilities than a persistent
notification. You should really all insist that apps make better
use of these.

Anyhow the current systray popup fits about 2-3 notifications.
It is not at all suited for serving as a notification history browser.
It would get overloaded so quickly that it would become useless
after about 30 minutes of desktop usage.

Finally, 
> The one thing KDE/PLASMA had it was that you could configure pretty much anything

You can't. It's a perception from the past where the user was
drowned in an ocean of options. This is what we're trying to get
away from, generally.

> I dont see whymaking a config should be the last resort.

Because it brings complexity on all possible fronts, including
maintenance, testing, bug handling, user psychology etc etc.

**Comment by colomar** [comment]
Configuration options are great if different people have different needs regarding an aspect. 
What we see too often in Plasma and KDE applications, however is the situation
- Version 1 (e.g. Plasma 4) did something in a way which wasn't really good
- Version 2 (e.g. Plasma 5) tried to improve the situation, but the solution wasn't really thought through
- People who found version 1 less of a pain than version 2 complain.
- We provide an option to switch between the two suboptimal versions.

This appeases those who complained about Version 2, but it leaves us with another config option and still no really satisfying solution. Sometimes, this is unavoidable. But sometimes, thinking a bit harder could have lead to a solution which serves everybody's needs.

Offering a config option is an "easy way out" which should not be out of the question, but should not be taken before thinking about a better solution.

In this case: Plasma 4 put all notifications in a history, which then became quickly so crowded that it was difficult to find the notification you were looking for in all the noise. Plasma 5 has a much more cleaner notification area, with the drawback that users may miss notifications which are important to them.

Before we offer users an option to just choose what's "the lesser of two evils" for them

**Comment by hetdegon** [comment]
Maybe some sort of whitelist based on the dbus id of the emitter? For example the notifications I care the most about are from pidgin and my (self-made*) download manager, while the rest are OK if they disappear after a bit. I bet everyone else has at least a couple programs they care the most about.

* the self-made note is not to brag, but to discourage stuff like "only a list of popular applications".

**Comment by mck182** [comment]
> my (self-made*) download manager

Since you're in control of your own notifications then, you can easily
make them persistent. If you're using KNotification, just set the flag.
If you're using the dbus api directly, you can set the timeout to 0.
This is exactly what you should be doing as an application developer.

This is also btw another good example for using KSNI (and also for
what it was invented for). With that you can have an icon in systray
that is Passive when nothing is being downloaded. Active when it
is downloading and NeedsAttention once it finishes downloading.
The icon can also indicate progress and display other useful data.
Its context menu can have full list of downloaded items with arbitrary
actions. It is more code, sure, but well worth it.

We do have the tech for "Notifications 2.0", we just need to use it.

**Comment by vindicatorr** [comment]
Frankly I am (mostly) pleased with Android's (kitkat) way of handling notifications, particularly with the slickdeals app.
Each app is a container within the notification container.
Slickdeals may have multiple notifications come in, but instead of having an individual swipe-dismissal for each slickdeal, they are all contained within the slickdeal container, so you can either click (to open) or swipe (to dismiss).

You could take it a bit further by taking other suggestions into play. Like limit the app container size to fit a certain # of notifications for that app.
Or have an app container display a # for the notification count for that app (program), along with the latest notification. The user can then click the notification where it will expand a certain distance and the user can scroll through the list, where specific sub notifications can be closed or the whole app container can be closed/dismissed.

I think I only have a handful of programs that submit notifications, so adding a scrollbar won't be too far reaching for such a small listing when each program is a container for it's own notifications.

**Comment by notmart** [comment]
@colomar an interesting thing would be the actual full text of her dissertation (even tough would be a looong read;) I guess depends from the rules there how much is freely distributable (for instance in Italy copyright goes to the university, but then goes on a license in which is freely consultable/readable, pretty much like any scientific paper not behind a paywall)

**Comment by colomar** [comment]
@notmart Here you go: http://contentdm.ad.umbc.edu/cdm/ref/collection/ETD/id/24842
And here is the gist of it condensed in a journal article: https://www.researchgate.net/publication/272373165_Interruptive_Notifications_in_Support_of_Task_Management
So choose your own adventure ;)

**Comment by notmart** [comment]
damn, I have to read it now /o\ (and i will, at least in part ;)

**Comment by christophschw** [comment]
> You can't. It's a perception from the past where the user was drowned in an ocean of options. This is what we're trying to get away from, generally.

Do we look like Mac-Users? ;)

**Comment by thugcee** [comment]
> It's a perception from the past where the user was drowned in an ocean of options. This is what we're trying to get away from, generally.

There is a lot of desktops for users who can't swim. Please leave this one for power users.

**Comment by christophschw** [comment]
> There is a lot of desktops for users who can't swim. Please leave this one for power users.

+1!

**Comment by nortexoid** [comment]
I don't know if people hadn't noticed, but people today are **more** proficient at using a computer, so more capable of understanding or avoiding configuration options. It doesn't make much sense to me to hear people (developers) arguing that we need to dumb down the user experience. What we need is a good balance given today's general proficiency level--which is pretty high!

**Comment by colomar** [comment]
An update on what others do:
- Unity: No history whatsoever
- Windows: No history whatsoever
- GNOME: They actually _do_ have a history, but no indicator for it. Instead, if you click on the time, you get a popover with, among other things - a history of notifications, regardless of whether they were permanent or not. So they do allow access to the history without annoying the user about it.

**Comment by christophschw** [comment]
If I'm not wrong: Unity has some Plugin to store all notifications.
And Windows (7) does not even have a notification Server. Nevermind...

Best would be if you could handle the behaviour on Server side. Which means the applications using libnotify would be registered and you could change the behaviour of each of them seperately: timeout, persistent flag, command execution...

**Comment by star-buck** [comment]
Windows: no history when they introduced a completely sophisticated
notification center in win10, that others already started copying like
deepin or budgie? I dont have win10, but even regardless of others, lets
bring back the functionality we already had in kde4.

**Comment by star-buck** [comment]
Win10 notifications area explained:
http://www.pcworld.com/article/2965932/windows/how-to-customize-what-appears-in-windows-10s-action-center.html

**Comment by mck182** [comment]
I would like to point out that windows10 uses a whole screen-height sidebar for notifications.

Plasma has very very limited place in form of the rather small systray popup, which would make notification history quite an unpleasant experience to use.

And I would like to once again reiterate to consider KSNIs instead. They are much better suited for everything everyone actually wants from notification history.

**Comment by colomar** [comment]
> Windows: no history when they introduced a completely sophisticated notification center in win10, that others already started copying like deepin or budgie?

Ah yes, sorry, I was only testing Windows 8.

**Comment by colomar** [comment]
> Plasma has very very limited place in form of the rather small systray popup, which would make notification history quite an unpleasant experience to use.

Good point, but I'm at a point where I feel (not only because of notifications) that we're limiting ourselves too much with that fixed-size popup, to be honest.

**Comment by mck182** [comment]
I don't disagree, here's a screenshot to illustrate it with 3 notifications: ![](http://paste.opensuse.org/view/raw/19adc51b)

**Comment by dbfin** [comment]
Just wanted to add one more vote for the feature I miss the most from KDE 4.

My use cases: Seafile telling me when some files had merging problems (this usually happens when I am not even near the computer, I figured it out accidently the last time), VPN being sometimes disconnected (Breeze icon is the same for wireless connection with or without VPN connected), Skype (where before I would set some events to have system notifications specifically because Skype's ones are not persistent), Email & Calendar (certain events), etc. etc. If I thought about it for another 5 minutes, I would easily recall another 5-10 use cases. In fact, I started realize how much I dependent on this persistent notifications feature when it disappeared.

I think the new system where applications decide which messages are important is good... theoretically. But in the best case scenario, where every single developer and every single OS finally appreciate and start using this specification, it will take at least another 5 years to adapt. And yet, after this, people will figure out that they still have different tastes. And something important for some is not so for others, and vice versa. They would start demanding "persistent notifications" option from each app and developer, to decide themselves on app-to-app basis.

So, wouldn't it just make more sense to support this feature for KDE eco-system as a whole, especially because KDE is the best (and only?) with the most integrated collection of common features among completely different apps? And this makes it a pleasant experience.

It seems, an easy way to do this would be to create a list. Similar to how it is done for icons in system tray, there could have been a list of apps ever produced any notifications, and for such apps there would be an option to set their future notifications as:
a) persistent regardless of what developers set (would be nice to see this as default)
b) auto according to what developers think about notification importance
c) non-persistent regardless of what developers set
d) banned (not even showing notifications from this app)

Not sure, if this is the best idea, but just wanted to point out that this feature is very important to me.

**Comment by star-buck** [comment]
That popup size limitation is not working well also for sound volume (thats
why we cant show individual streams), for network info, etc.

Besides that, at least for notifications, the notes could stack or we could
think of a smart way? On a phone, it is evident that a lot of infos about
messenges, mail, calls etc. stack up regularly and need to be cleared all
at once or individually.

For the size, notice that if part of the systray, the size is limited and
equal to all others, but if you individually dragndrop an applet in the
panel, like volume or notifications, the size is already somewhat different
and could be freely adjustable or expanded.

**Comment by christophschw** [comment]
> It seems, an easy way to do this would be to create a list. Similar to how it is done for icons in system tray, there could have been a list of apps ever produced any notifications, and for such apps there would be an option to set their future notifications as:
a) persistent regardless of what developers set (would be nice to see this as default)
b) auto according to what developers think about notification importance
c) non-persistent regardless of what developers set
d) banned (not even showing notifications from this app)

+1 for this!

**Comment by thugcee** [comment]
@dbfin +1

**Comment by christophschw** [comment]
Any progress? Thanks

**Comment by Govikk** [comment]
The other problem with non-persistent notification is that you can't copy message to clipboard. 
E.g. you have long error from network manager and you can not to copy to check or to google it.

**Comment by kokoko3k** [comment]
I'm still on kde4 and this keeps me from switching to plasma 5.
An idea: what about a new plasmoid that keeps the notification history so that who wants the history would simply add it to the workspace?

Thanks for your efforts, devs.

**Comment by heikojansen** [comment]
The disappearing notifications - especially (but not only) those from KMail - are one the most annoying aspects of KDE/Plasma for me, too.
KMail apparently does not offer any configuration so that it would mark it´s notifications as persistent and I suppose that´s also the case for other applications. 
So while I´m going to file a bug report with KMail in a minute (or add my comment to an hopefully existing one) I´d really appreciate it if a fallback option to keep certain kinds of notifications could be added to the plasma notification component...

**Comment by maces** [comment]
@kokoko3k generally like the idea. 
I would prefer something not bound to the desktop, as I relay see it, but rather like an expandable popup or a sidebar like on mac (here i rarely miss a notification):
- chat messages are displayed even on the lock screen (debatable because of privacy, on android you can configure this like: no notifications, only title, full message + options)
- swipe from the right displays the right sidebar which has enough space to easily manage many notifications (or a wider popup like in gnome, resulting in less new lines)

so a split in the kde notification server like @kokoko3k described could leave the notification popup to display only the most recent ones and a button for more which would open the **sidebar/...** (also reachable by keys/gesture). This sidebar could be totally optional.
I'm not quite getting why the popup is so restricted on height. Could it not be expanded based on available screen space? A 'more' button would be totally usable also.


The **filtering** suggested by @dbfin would be the ideal way, imho. Only change defaulting to the developers intended setting/behave like now. 
adding to @vindicatorr form 2nd feb:
Especially with many software (even commercially one) being good at something but lack those notification features due to no will or no update, this could really help managing them.


Additionally once implemented, KDE could have a **'no interruptions'-mode** like on android. This could maybe address some of your concerns regarding usability @colomar ? Or at least propose another use case:
On android the notifications work rather well, so I can use them as a very basic TODO-reminder only removing those TODOs which are done aggregated through all applications. One quick look vs checking in Mail, Messenger, X, Y, Z, making me much faster! 
Combined with a  'no interruptions'-quick-option one could work with less interruptions but not miss on any of them. (Could be integrated with KDEConnect later on, so  calls are also not interrupting ect.)

TL:DR;
- sidebar to gain space to manage a notification history
- filtering to master all applications behaviour and own preference
- 'no interruptions'-mode for work sessions demanding that extra bit of concentration

**Comment by kokoko3k** [comment]
If you don't mind using gambas tools (some people do for unknown reasons), then check:
https://github.com/kokoko3k/permanotices
It is a daemon-like tool that aims to replace notifications with persistent ones.

**Comment by kokoko3k** [comment]
"colomar was unassigned by kokoko3k 3 days ago"
What does it means?

**Comment by star-buck** [comment]
good question :)
As for colomar, he has a different day-to-day job and therefore isn't going to resolve this anyway, so thats fine. Also there are some plans for revising part or the whole notification thing, as its ancient and not really up-to-date. We will see.

**Comment by kokoko3k** [comment]
Oh thanks,
It is just that i was puzzled by the "unassigned by kokoko3k" part, as i did not expicitely unassigned anyone :)

**Comment by star-buck** [comment]
Continuing the idea here:
https://github.com/blue-systems/plasma-ideas/issues/35

**Comment by UralZima** [comment]
Please support this here:  https://bugs.kde.org/show_bug.cgi?id=378032

**Comment by rthardy** [comment]
This one needs my two cents, since I've spent a couple hours looking into it and reading this forum.

I often notice a pop-up with buttons on it, just as it disappears.  Was it important?  I can't say, but apparently the developer didn't think so.  Can I refer to the application for a clue?  No, because I have no idea what it was about or what application produced it.  This is annoying.

Meanwhile I have yet to see a persistent notification after a year with KF5.  The systray notification icon is apparently there to take up space and get me to spend time looking into the apparent problem.

This could use a little work.

**Comment by UralZima** [comment]
@rthardy, please double your post here: https://bugs.kde.org/show_bug.cgi?id=378032
We are actively discussing this with developers and your post will be useful.

**Comment by rthardy** [comment]
Done.

**Comment by JackDinn** [comment]
+1 yes very annoying that i have regressed in this regard since installing plasma 5. But as usual when i look up why things are missing etc i find a bug thread thats years and years long without progress :(

**Comment by UralZima** [comment]
this is not official bug tracker for plasma project. It is just for bluesystem's needs. Please make your vote here: https://bugs.kde.org/show_bug.cgi?id=378032

**Comment by intika** [comment]
I made an applet that keep all the notifications based on the current one, awaiting KDE 5.11 to become stable and widely distributed 
https://store.kde.org/p/1227099/
https://github.com/intika/org.kde.plasma.notifications.full
